pkg/descheduler/metricscollector/metricscollector.go

```diff
@@ -0,0 +1,144 @@
+	}, 5*time.Second, ctx.Done())
+}
+
+func weightedAverage(prevValue, value int64) int64 {
+	return int64(math.Round(beta*float64(prevValue) + (1-beta)*float64(value)))
+}
+
+func (mc *MetricsCollector) AllNodesUsage() (map[string]map[v1.ResourceName]*resource.Quantity, error) {
+	mc.mu.RLock()
+	defer mc.mu.RUnlock()
+
+	allNodesUsage := make(map[string]map[v1.ResourceName]*resource.Quantity)
+	for nodeName := range mc.nodes {
+		allNodesUsage[nodeName] = map[v1.ResourceName]*resource.Quantity{
+			v1.ResourceCPU:    utilptr.To[resource.Quantity](mc.nodes[nodeName][v1.ResourceCPU].DeepCopy()),
+			v1.ResourceMemory: utilptr.To[resource.Quantity](mc.nodes[nodeName][v1.ResourceMemory].DeepCopy()),
+		}
+	}
+
+	return allNodesUsage, nil
+}
+
+func (mc *MetricsCollector) NodeUsage(node *v1.Node) (map[v1.ResourceName]*resource.Quantity, error) {
+	mc.mu.RLock()
+	defer mc.mu.RUnlock()
+
+	if _, exists := mc.nodes[node.Name]; !exists {
+		klog.V(4).InfoS("unable to find node in the collected metrics", "node", node.Name)
+		return nil, fmt.Errorf("unable to find node %q in the collected metrics", node.Name)
+	}
+	return map[v1.ResourceName]*resource.Quantity{
+		v1.ResourceCPU:    utilptr.To[resource.Quantity](mc.nodes[node.Name][v1.ResourceCPU].DeepCopy()),
+		v1.ResourceMemory: utilptr.To[resource.Quantity](mc.nodes[node.Name][v1.ResourceMemory].DeepCopy()),
+	}, nil
+}
+
+func (mc *MetricsCollector) HasSynced() bool {
+	return mc.hasSynced
+}
+
+func (mc *MetricsCollector) MetricsClient() metricsclient.Interface {
+	return mc.metricsClientset
+}
+
+func (mc *MetricsCollector) Collect(ctx context.Context) error {
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+	nodes, err := mc.nodeLister.List(context.TODO(), metav1.ListOptions{LabelSelector: mc.nodeSelector})
+	if err != nil {
+		return fmt.Errorf("unable to list nodes: %v", err)
+	}
+
+	for _, node := range nodes.Items {
+		metrics, err := mc.metricsClientset.MetricsV1beta1().NodeMetricses().Get(ctx, node.Name, metav1.GetOptions{})
+		if err != nil {
+			klog.Errorf("Error fetching metrics for node %s: %v\n", node.Name, err)
+			// No entry -> duplicate the previous value -> do nothing as beta*PV + (1-beta)*PV = PV
+			continue
+		}
+
+		if _, exists := mc.nodes[node.Name]; !exists {
+			mc.nodes[node.Name] = map[v1.ResourceName]*resource.Quantity{
+				v1.ResourceCPU:    utilptr.To[resource.Quantity](metrics.Usage.Cpu().DeepCopy()),
+				v1.ResourceMemory: utilptr.To[resource.Quantity](metrics.Usage.Memory().DeepCopy()),
+			}
+		} else {
+			// get MilliValue to reduce loss of precision
+			mc.nodes[node.Name][v1.ResourceCPU].SetScaled(
+				weightedAverage(mc.nodes[node.Name][v1.ResourceCPU].ScaledValue(resource.Micro), metrics.Usage.Cpu().ScaledValue(resource.Micro)),
+				resource.Micro,
+			)
+			mc.nodes[node.Name][v1.ResourceMemory].SetScaled(
+				weightedAverage(mc.nodes[node.Name][v1.ResourceMemory].ScaledValue(resource.Micro), metrics.Usage.Memory().ScaledValue(resource.Micro)),
```

Comment:
This seems a bit dangerous for the memory. We should make sure we do not cause overflow. Isn't just calling Value enough here?

Reply:
Good point.

Reply:
Done